### DISCUSSION TITLE
Discuss solution: Sync child applications from parent application honoring sync-waves

### DISCUSSION THREAD

**Discussion by mattesja** [argoproj/argo-cd#18416] [Q&A]
We have the same requirement, which is described in this [Bug Report ](https://github.com/argoproj/argo-cd/issues/8778#issuecomment-1069168965):

- We are using app of apps
- Child applications have different sync waves
- After updating the image tags of the child applications, the child applications should be deployed in the order of the sync waves.

I've learned, that ArgoCD does not support this requirement: https://github.com/argoproj/argo-cd/issues/8778#issuecomment-1069168965

Therefore I've worked out a solution, which will trigger the sync of the child applications from the parent application in the correct sync waves order:

- I've introduced sync-hooks fro each application at the app-of-apps level
- Once the parent application is synced, the hooks will trigger a sync of the child applications via REST api in the order of the sync waves.
- The hook also wait for the apps's health.

You will find all details here: https://github.com/mattesja/argocd-demo

**Discussion**

- Is there a better solution to achieve this?
- This solution is not thread safe: The app might not yet be restarted, when testing the health.
- Parsing json via sed is not a good idea. But I wanted to used standard docker images in that demo.

**Sync Script**

```
apiVersion: batch/v1
kind: Job
metadata:
  generateName: sync-child-application-app-1
  annotations:
    argocd.argoproj.io/hook: PostSync
    argocd.argoproj.io/hook-delete-policy: HookSucceeded
    argocd.argoproj.io/sync-wave: "1"
spec:
  template:
    spec:
      containers:
        - name: sync-child-application
          image: curlimages/curl
          command:
            - "/bin/sh"
            - "-c"
            - |
              ADMIN_PW=$(cat /etc/secret/password)
              echo "retrieve token"
              ARGOCD_TOKEN=$(curl -ss -k -X POST -H "Content-Type: application/json" https://argo-argocd-server.argocd.svc.cluster.local/api/v1/session -d "{\"username\":\"admin\",\"password\":\"$ADMIN_PW\"}"|sed "s/.*token.*\"\(.*\)\".*/\1/")

              syncstatus=$(curl -ss -k -H "Content-Type: application/json"  https://argo-argocd-server.argocd.svc.cluster.local/api/v1/applications/app-1  -H "Authorization: Bearer $ARGOCD_TOKEN"| sed -n 's/.*"sync":{"status":"\([^"]*\)".*/\1/p')
              echo "syncstatus: $syncstatus"
              if [ "$syncstatus" == "Synced" ]; then
                echo "Application is synced. Exiting..."
                exit 0
              fi

              echo "sync app"
              curl -ss -k -X POST -H "Content-Type: application/json"  https://argo-argocd-server.argocd.svc.cluster.local/api/v1/applications/app-1/sync -H "Authorization: Bearer $ARGOCD_TOKEN"

              echo "wait for health"
              # Trigger sync and then waiting for health is not thread safe!
              # We might encounter a race condition where the app health is still healthy, because the app was not restarted yet.
              sleep 30
              status=""
              while [ "$status" != "Healthy" ]; do
                sleep 5
                echo "waiting for app-1 to be healthy"
                status=$(curl -ss -k -H "Content-Type: application/json"  https://argo-argocd-server.argocd.svc.cluster.local/api/v1/applications/app-1  -H "Authorization: Bearer $ARGOCD_TOKEN" | sed -n 's/.*"health":{"status":"\([^"]*\)".*/\1/p')
              done
              echo "app-1 is healthy"
          volumeMounts:
            - name: secret-volume
              mountPath: /etc/secret
              readOnly: true
      volumes:
        - name: secret-volume
          secret:
            secretName: argocd-initial-admin-secret
      restartPolicy: Never
  backoffLimit: 2
---
...
 ```

**References and My Research**

There are some documentation and discussions, of how to use sync waves in the app-of-apps pattern.

But this solutions only work on creation time:
> So if your use-case is to sync those apps in a specific order at bootstrap time (e.g. when they're initially created), this will work fine.
https://github.com/argoproj/argo-cd/discussions/6447

Updating image tags in sync waves of parent applications is not supported:
>  In the Apps-of-Apps pattern the only resource managed by the parent app are the child Applications themselves. Image tags are not part of the Application resource and so are not reported in diffs. ArgoCD will always compare the desired state with the live state and in the apps-of-apps case, only the Application resource changes are validated.
https://github.com/argoproj/argo-cd/issues/8778#issuecomment-1069168965

- https://codefresh.io/blog/argo-cd-application-dependencies/ Blog post of how to use sync waves in app-of-apps pattern
- https://github.com/argoproj/argo-cd/discussions/18230 Show sync status at parent level

**Comment by ronaldour** [comment]
could this be achieved creating a [custom resource action](https://argo-cd.readthedocs.io/en/stable/operator-manual/resource_actions/) using LUA? 🤔
